internal/driver/common/canvas_test.go

```diff
@@ -0,0 +1,375 @@
+package common
+
+import (
+	"image/color"
+	"testing"
+
+	"fyne.io/fyne/v2"
+	"fyne.io/fyne/v2/canvas"
+	"fyne.io/fyne/v2/container"
+	"fyne.io/fyne/v2/test"
+	"fyne.io/fyne/v2/theme"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestCanvas_walkTree(t *testing.T) {
+	test.NewApp()
+	defer test.NewApp()
+
+	leftObj1 := canvas.NewRectangle(color.Gray16{Y: 1})
+	leftObj2 := canvas.NewRectangle(color.Gray16{Y: 2})
+	leftCol := container.NewWithoutLayout(leftObj1, leftObj2)
+	rightObj1 := canvas.NewRectangle(color.Gray16{Y: 10})
+	rightObj2 := canvas.NewRectangle(color.Gray16{Y: 20})
+	rightCol := container.NewWithoutLayout(rightObj1, rightObj2)
+	content := container.NewWithoutLayout(leftCol, rightCol)
+	content.Move(fyne.NewPos(17, 42))
+	leftCol.Move(fyne.NewPos(300, 400))
+	leftObj1.Move(fyne.NewPos(1, 2))
+	leftObj2.Move(fyne.NewPos(20, 30))
+	rightObj1.Move(fyne.NewPos(500, 600))
+	rightObj2.Move(fyne.NewPos(60, 70))
+	rightCol.Move(fyne.NewPos(7, 8))
+
+	tree := &renderCacheTree{root: &RenderCacheNode{obj: content}}
+	c := &Canvas{}
+	c.Initialize(nil, 4096, func() {})
+	c.SetContentTreeAndFocusMgr(&canvas.Rectangle{FillColor: theme.BackgroundColor()})
+
+	type nodeInfo struct {
+		obj                                     fyne.CanvasObject
+		lastBeforeCallIndex, lastAfterCallIndex int
+	}
+	updateInfoBefore := func(node *RenderCacheNode, index int) {
+		pd, _ := node.painterData.(nodeInfo)
+		if (pd != nodeInfo{}) && pd.obj != node.obj {
+			panic("node cache does not match node obj - nodes should not be reused for different objects")
+		}
+		pd.obj = node.obj
+		pd.lastBeforeCallIndex = index
+		node.painterData = pd
+	}
+	updateInfoAfter := func(node *RenderCacheNode, index int) {
+		pd := node.painterData.(nodeInfo)
+		if pd.obj != node.obj {
+			panic("node cache does not match node obj - nodes should not be reused for different objects")
+		}
+		pd.lastAfterCallIndex = index
+		node.painterData = pd
+	}
+
+	//
+	// test that first walk calls the hooks correctly
+	//
```

Comment:
Suspicious comment formatting?

Reply:
Not sure, it seems that the comment was written like that to clearly divide the tests?